### PR TITLE
Updated 32 bit heuristics

### PR DESCRIPTION
**Pull request by chflood**
We try really hard to stay under 2gb but if the live data exceeds that
size we keep the interval small and perform full gcs until we get back
under 2gb.

### CODE REVIEW

File: src/gc.c [hunk 1]
```diff
@@ -585,12 +585,20 @@ static void gc_sweep_foreign_objs(void)
 // GC knobs and self-measurement variables
 static int64_t last_gc_total_bytes = 0;
 
+// max_total_memory is a suggestion.  We try very hard to stay
+// under this limit, but we will go above it rather than halting.
 #ifdef _P64
 #define default_collect_interval (5600*1024*sizeof(void*))
 static size_t max_collect_interval = 1250000000UL;
+// Eventually we can expose this to the user/ci.
+static uint64_t max_total_memory = (uint64_t) 2 * 1024 * 1024 * 1024 * 1024 * 1024;
```
Comment (oscardssmith) [inline src/gc.c:594]: 2 exabytes? That seems a little excessive...

Comment (chflood) [reply src/gc.c:594]: The idea is that the 64 bit heuristics will be limited by the available memory and the 32 bit heuristics are limited by the available address space.  The 2 exabytes was arbitrary but hopefully large enough to not be limiting.

**Review by oscardssmith — APPROVED** [review]

**Comment by ViralBShah** [comment]
Is this going to fix our linux32 and win32 build failures? Looks promising - at least win32 is passing.

**Comment by giordano** [comment]
What happened to the fixed debug print statements?

**Comment by oscardssmith** [comment]
I believe they got fixed in a separate PR.

**Comment by giordano** [comment]
I don't think so.  They are mentioned in the title of this pull request, they were introduced in https://github.com/JuliaLang/julia/pull/44805/commits/b82679a056a8c724ba3e0a32f0adcc5c005013f2.  Af far as I can tell, currently in `master` these lines were last touched by me: https://github.com/JuliaLang/julia/blame/f536b81aea93af0ed03543198560cf436e7def51/src/gc-debug.c#L985

**Comment by chflood** [comment]
I wanted to separate the two issues because I thought you guys were fixing
this.

I think the better way to fix it is to use PRI macros and I can put
together a PR.

Christine


Christine

**Comment by giordano** [comment]
Having the fix in a different pull request is totally fine (and probably even better, as it's slightly orthogonal to this pull request), but that's still mentioned in the title of this PR and I don't see any other PR with it. Hence why I'm wondering what happened to it 🙂

**Comment by chflood** [comment]
OK, I just submitted the fix for the print errors in a separate PR.

**Comment by oscardssmith** [comment]
Ready to merge?

**Comment by chflood** [comment]
I think so.

**Comment by DilumAluthge** [comment]
Thanks @chflood!

**Pull request merged**